### PR TITLE
Add Optimization Cookbook

### PR DESCRIPTION
**Pull request by samanklesaria**
# What does this PR do?

This PR adds a guide that shows some common techniques for working with Flax models during optimization. These include:

- Calculation of Exponential Moving Averages
- Optimizing only a low rank addition to certain weights (LORA)
- Using different learning rates for different parameters to implement the maximal update parameterization
- Using second order optimizers like LBFGS. 
- Specifying sharding for optimization state that differs from that of parameter state
- Gradient accumulation

This document emphasizes a style as close to pure jax as possible: to that end, it shows how the flax version of each technique only requires minor deviation from the often more intuitive pure-jax version. 

# Warnings:

- This PR requires the fix #5171 , which is why the diff looks so big here. This PR shouldn't be merged until after #5171 
- This PR includes both the notebook file used to construct the optimization examples as well as the RST file that contains what gets used in the documentation. When review is complete (and I don't need to actually test any of the python snippets anymore) I will delete the notebook file.

### CODE REVIEW

**Comment by cgarciae** [comment]
I feel we could simplify the intro by doing the following:

1. Define a single model at the begining, simply reuse it on all examples (its just a guide).
2. Inline the loss function.

```python
model = nnx.Sequential(
  nnx.Linear(2,8, rngs=rngs),
  nnx.relu,
  nnx.Linear(8,8, rngs=rngs),
)

optimizer = nnx.Optimizer(
  model,
  tx=optax.adam(1e-3),
  wrt=nnx.Param)

...

@nnx.jit
def train_step(model, optimizer, ema, x, y):
  loss_fn = lambda m, x, y: jnp.sum((m(x) - y) ** 2)
  loss, grads = nnx.value_and_grad(loss_fn)(model, x, y)
  optimizer.update(model, grads)
  ema.update(model)
  return loss
```

File: docs_nnx/guides/opt_cookbook.rst [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+      wrt=nnx.Param)
+
+    def ema_update(ema, new_val, decay=0.9):
+        return decay * ema + (1 - decay) * new_val
+
+    class Ema(nnx.Module):
+        def __init__(self, model):
+            self.ema = nnx.clone(model) # Make a copy
+        def update(self, model):
+            self.ema = tree.map(ema_update, self.ema, model)
```
Comment (cgarciae) [inline docs_nnx/guides/opt_cookbook.rst]:
```suggestion
    class Ema(nnx.Module):
        def __init__(self, model, decay=0.9):
            self.decay = decay
            self.ema = jax.tree.map(jnp.copy, model) # Make a copy
        def update(self, model):
            def ema_update(ema, new_val):
              return decay * ema + (1 - self.decay) * new_val
            self.ema = tree.map(ema_update, self.ema, model)
```

File: docs_nnx/guides/opt_cookbook.rst [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+    :title: NNX, Jax
+    :sync:
+
+    def add_rank2_lora(path, node):
+      if isinstance(node, nnx.Linear):
+        return nnx.LoRA(node.in_features, 2, node.out_features, base_module=node, rngs=rngs)
+      return node
+
+    base_model = nnx_model(rngs)
+    model = nnx.recursive_map(add_rank2_lora, base_model)
```
Comment (cgarciae) [inline docs_nnx/guides/opt_cookbook.rst]: would be nice to `nnx.display(model)` or print here. I guess RST is a bit of a limitation.

File: docs_nnx/guides/opt_cookbook.rst [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+          value=loss,
+          value_fn=loss_fn_state)
+        return loss
+
+    model = nnx_model(rngs)
+
+    nnx_optimizer = nnx.Optimizer(
+      model,
+      tx=optax.lbfgs(1e-3),
+      wrt=nnx.Param)
```
Comment (cgarciae) [inline docs_nnx/guides/opt_cookbook.rst]:
```suggestion
      wrt=nnx.LoRAParam)
```

Comment (cgarciae) [reply docs_nnx/guides/opt_cookbook.rst]: We also need to explain that we are calculating the gradients for all the parameters in `grad` but are filtering them here and it might be ok because of DCE.

File: docs_nnx/guides/opt_cookbook.rst [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+In Flax, we will also initialize a partitioned optax optimizer. But unlike the Jax case where we use the model as the PyTree, we'll need to use the model's *state* PyTree here instead. These state PyTrees aren't compatible with ordinary jax tree mapping; instead of ``jax.tree.map_with_path``, we'll need to use the NNX version, which is called ``nnx.map_state``.
+
+.. codediff::
+    :title: NNX, Jax
+    :sync:
+
+    model = nnx_model(rngs)
+    state = nnx.state(model, nnx.Param)
+    rates = {'kernel': optax.adam(1e-3), 'bias': optax.adam(1e-2)}
+    param_tys = nnx.map_state(lambda p, v: list(p)[-1], state)
```
Comment (cgarciae) [inline docs_nnx/guides/opt_cookbook.rst]: I thin this is enough:
```suggestion
    param_tys = nnx.map_state(lambda p, v: p[-1], state)
```

Comment (cgarciae) [reply docs_nnx/guides/opt_cookbook.rst]: We could also use `jax.tree. map_with_path` as in the JAX example.

File: docs_nnx/guides/opt_cookbook.rst [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+
+.. codediff::
+    :title: NNX, Jax
+    :sync:
+
+    mesh = jax.make_mesh((2, 4), ("x", "y"),
+                         axis_types=(AxisType.Explicit, AxisType.Explicit))
+    jax.set_mesh(mesh)
+
+    ghost_model = jax.eval_shape(lambda: nnx_model(nnx.Rngs(0), out_sharding=P('x', 'y')))
```
Comment (cgarciae) [inline docs_nnx/guides/opt_cookbook.rst]: instead of creating this fake model it would be a good opportunity to create the `optimizer_sharding` API on Variable before finishing this guide.

File: flax/nnx/helpers.py [hunk 6]
```diff
@@ -185,6 +186,17 @@ def __setitem__(self, index: int | slice, value: A | tp.Iterable[A]) -> None:
     else:
       raise TypeError('Invalid index type')
 
+  def _graph_node_set_key(self, key: str, value: tp.Any):
```
Comment (cgarciae) [inline flax/nnx/helpers.py:198]: Is this still needed?

**Comment by cgarciae** [comment]
After fully reading the guide I'm getting the sense that having the JAX versions makes explanations a bit longer and slightly harder to understand (cause you have to mentally filter for the version you are interested in) and having the JAX version doesn't necessarily make understanding the NNX version easier.

**Comment by samanklesaria** [comment]
> After fully reading the guide I'm getting the sense that having the JAX versions makes explanations a bit longer and slightly harder to understand (cause you have to mentally filter for the version you are interested in) and having the JAX version doesn't necessarily make understanding the NNX version easier.

Fair enough! I'll convert it to nnx-only.

File: flax/nnx/helpers.py (hunk 6 above)
Comment (samanklesaria) [reply flax/nnx/helpers.py:198]: I don't think so. Removing it.

File: docs_nnx/guides/opt_cookbook.rst (hunk 2 above)
Comment (samanklesaria) [reply docs_nnx/guides/opt_cookbook.rst]: Done.

File: docs_nnx/guides/optimization_cookbook.md [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+
+rngs = nnx.Rngs(0)
+
+def make_model(rngs, **metadata):
+    return nnx.Sequential(
+        nnx.Linear(2,8, rngs=rngs, kernel_metadata=metadata),
+        nnx.Linear(8,8, rngs=rngs, kernel_metadata=metadata))
+
+def loss_fn(model, x, y):
+    return jnp.sum((model(x) - y) ** 2)
```
Comment (cgarciae) [inline docs_nnx/guides/optimization_cookbook.md]: should we use `mean` here?

File: docs_nnx/guides/optimization_cookbook.md [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+```
+
+# Exponential Moving Average
+
+Neural network see increased robustness when, rather than using only the weights available at the end of training, we use an exponential moving average of the weights produced throughout training. To modify the standard Flax training loop to accomodate calculating exponential moving averages, we can introduce a new `nnx.Variable` subclass for EMA Params, along with a function that converts all variables in a module to this subclass. 
+
+```python
+class EmaParam(nnx.Variable):
+  @classmethod
+  def from_variable(cls, var):
```
Comment (cgarciae) [inline docs_nnx/guides/optimization_cookbook.md:54]: Seems like this classmethod could just be a function in `as_ema_params`?

File: docs_nnx/guides/optimization_cookbook.md [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+
+```python
+# initialization
+model = make_model(rngs)
+ema = EMA(model, decay=0.9)
+
+# simulate parameter update
+def double(param):
+  param[...] *= 2.0
+jax.tree.map(double, model, is_leaf=lambda x: isinstance(x, nnx.Variable))
```
Comment (cgarciae) [inline docs_nnx/guides/optimization_cookbook.md:98]: we could use the new `nnx.map` here.
